community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,112 +19,302 @@
- *     under synchronisation.
+ * In a sense, this turns many small, presumably concurrent, pieces of work
+ * into fewer, larger batches of work, that is then applied to the material
+ * under synchronisation.
  * </p>
  * <p>
- *     Obviously this only makes sense for work that a) can be combined, and b)
- *     where the performance improvements from batching effects is large enough
- *     to overcome the overhead of collecting and batching up the work units.
+ * Obviously this only makes sense for work that a) can be combined, and b)
+ * where the performance improvements from batching effects is large enough
+ * to overcome the overhead of collecting and batching up the work units.
  * </p>
+ *
  * @see Work
  */
 public class WorkSync<Material, W extends Work<Material,W>>
 {
     private final Material material;
     private final AtomicReference<WorkUnit<Material,W>> stack;
     private final WorkUnit<Material,W> stackEnd;
-    private final ReentrantLock lock;
+    private final AtomicReference<Thread> lock;
 
     /**
      * Create a new WorkSync that will synchronize the application of work to
      * the given material.
+     *
      * @param material The material we want to apply work to, in a thread-safe
      * way.
      */
     public WorkSync( Material material )
     {
         this.material = material;
-        this.stackEnd = new WorkUnit<>( null );
+        this.stackEnd = new WorkUnit<>( null, null, null );
+        this.stackEnd.complete();
         this.stack = new AtomicReference<>( stackEnd );
-        this.lock = new ReentrantLock();
+        this.lock = new AtomicReference<>();
     }
 
     /**
      * Apply the given work to the material in a thread-safe way, possibly by
      * combining it with other work.
+     *
      * @param work The work to be done.
+     * @throws ExecutionException if this thread ends up performing the piled up work,
+     * and any work unit in the pile throws an exception. Thus the current thread is not
+     * gauranteed to observe any exception its unit of work might throw, since the
+     * exception will be thrown in whichever thread that ends up actually performing the work.
      */
-    public void apply( W work )
+    public void apply( W work ) throws ExecutionException
     {
         // Schedule our work on the stack.
-        WorkUnit<Material,W> unit = new WorkUnit<>( work );
-        unit.next = stack.getAndSet( unit ); // benign race, see reverse()
+        WorkUnit<Material,W> unit = enqueueWork( work );
 
         // Try grabbing the lock to do all the work, until our work unit
         // has been completed.
-        boolean wasInterrupted = false;
         int tryCount = 0;
         do
         {
             tryCount++;
-            try
+            checkFailure( tryDoWork( unit, tryCount, true ) );
+        }
+        while ( !unit.isDone() );
+    }
+
+    /**
+     * Apply the given work to the material in a thread-safe way, possibly asynchronously
```

Comment:
The _"possibly asynchronously"_ part here differs from the contract implied by `Future`:

> A `Future` represents the result of an asynchronous computation.

It only _possibly_ being asynchronous is a slight deviation from that contract. For a "normal" `Future` it would be possible to poll `isDone()` occasionally, and never call the `get()` method until it is known that it will not block. With this implementation `get()` must be called (by at least one of the involved threads) in order for the work to be done.